### PR TITLE
[TACHYON-639] Avoid copying ByteBuf data to DataByteBuffer in RPCBlockResponse#decode()

### PR DESCRIPTION
**Pull request by ooq**
https://tachyon.atlassian.net/browse/TACHYON-639

Turns out we can avoid copying the data from ByteBuf to DataByteBuffer easily: simply pass the reference of the underlying ByteBuffer. Even though `ByteBuf` will be deallocated soon after decoding, the underlying ByteBuffer stays as its reference counting increases when we passed it to DataByteBuffer.

I'll see if I can benchmark the benefit of this patch.

@gpang

### CODE REVIEW

**Comment by ooq** [comment]
```
  @Test
  public void encodeDecodePerfTest() {
    int length = 10 * 1000 * 1000;
    DataByteBuffer payload = new DataByteBuffer(ByteBuffer.allocate(length), length);
    RPCBlockResponse resp = new RPCBlockResponse(BLOCK_ID, OFFSET, length, payload, STATUS);
    resp.encode(mBuffer);
    mBuffer.writeBytes(payload.getReadOnlyByteBuffer());
    long startTime = System.nanoTime();
    RPCBlockResponse resp2 = RPCBlockResponse.decode(mBuffer);
    long endTime = System.nanoTime();
    System.out.print("Time elasped: " + (endTime - startTime) / 1000 + "ms.");
    assertValid(BLOCK_ID, OFFSET, length, STATUS, resp);
    assertValid(BLOCK_ID, OFFSET, length, STATUS, resp2);
  }
```

For the test above, the console output would be 80ms (no copy) vs. 13000ms (before).

**Comment by gpang** [comment]
Thanks for looking into this @ooq ! Is the test result supposed to be 13,000 ms or 1,300 ms? It takes over a second to copy 10 mb?

**Comment by ooq** [comment]
@gpang 13000ms. Can you try the test and let me know your numbers?

Thanks.

**Comment by ooq** [comment]
sorry...80us vs. 13ms

**Comment by haoyuan** [comment]
We can’t automatically merge this pull request.

**Comment by ooq** [comment]
Jenkins, test this please.

**Comment by ooq** [comment]
@gpang @apc999 @calvinjia @haoyuan Can some of you review this PR? Thanks.
To summarize: a new class `DataNettyBuffer`, which wraps up a `ByteBuf` (as opposed to copy it when we used `DataByteBuffer`). To avoid the `ByteBuf` to be recycled by Netty, we call `ByteBuf#retain()` during the construction and `ByteBuf#release()` when reading is finished.

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockInStream.java [hunk 1]
```diff
@@ -103,6 +104,9 @@
    */
   private static final int MAX_REMOTE_READ_ATTEMPTS = 2;
 
+  /** A reference to the current reader so we can clear it after reading is finished. */
+  private static RemoteBlockReader sUnclearedReader = null;
```
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockInStream.java]: I don't think this should be 'static'.

Also, maybe naming it as "mCurrentReader" would make more sense.

File: clients/unshaded/src/main/java/tachyon/client/netty/NettyRemoteBlockReader.java [hunk 2]
```diff
@@ -88,4 +93,18 @@ public ByteBuffer readRemoteBlock(String host, int port, long blockId, long offs
       throw new IOException(e);
     }
   }
+
+  /**
+   * Clear the previous read response, release the resource the response references.
+   *
+   * @return true if the response is cleared, or there is nothing needs to be cleared.
+   */
+  public boolean clearReadResponse() {
```
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/netty/NettyRemoteBlockReader.java]: I think it would be nice if the "RemoteBlockReader" interface included a close() method, so other classes just have to close() this reader when desired. Maybe "RemoteBlockReader" should extend "Closeable"?

File: common/src/main/java/tachyon/network/protocol/RPCBlockReadResponse.java [hunk 3]
```diff
@@ -128,4 +125,16 @@ public long getOffset() {
   public Status getStatus() {
     return mStatus;
   }
+
+  /**
+   * Release the underlying DataBuffer in this read response.
+   *
+   * @return true if successfully released or no release is needed; otherwise false.
+   */
+  public boolean releaseBuffer() {
```
Comment (gpang) [inline common/src/main/java/tachyon/network/protocol/RPCBlockReadResponse.java]: For releasing the DataBuffer, I think the other classes can just call "getPayloadDataBuffer()" to get the DataBuffer, and then call release() on the data buffer when it is done. This would mean that the DataBuffer interface would need a release() method.

This would also mean there should be a comment on the getPayloadDataBuffer() interface that release() must be called on it.

I just didn't think that release() or close() on a message object made as much sense as releasing a DataBuffer. What do you think?

File: common/src/main/java/tachyon/network/protocol/databuffer/DataNettyBuffer.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+/**
+ * A DataBuffer with the underlying data being a {@link ByteBuf}.
+ */
+public class DataNettyBuffer extends DataBuffer {
+  private final ByteBuffer mBuffer;
+  private final long mLength;
+  private ByteBuf mNettyBuf = null;
```
Comment (gpang) [inline common/src/main/java/tachyon/network/protocol/databuffer/DataNettyBuffer.java]: Can this be final like the other variables?

File: common/src/main/java/tachyon/network/protocol/databuffer/DataNettyBuffer.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    // increase the bytebuf reference count so it would not be recycled by Netty
+    bytebuf.retain();
+    mNettyBuf = bytebuf;
+    mBuffer = bytebuf.nioBuffer();
+    mLength = length;
+  }
+
+  @Override
+  public Object getNettyOutput() {
+    return Unpooled.wrappedBuffer(mBuffer);
```
Comment (gpang) [inline common/src/main/java/tachyon/network/protocol/databuffer/DataNettyBuffer.java]: I think you can just return mBuffer here, since it is already a ByteBuf (a supported Netty output class).

File: common/src/main/java/tachyon/network/protocol/databuffer/DataNettyBuffer.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+  @Override
+  public ByteBuffer getReadOnlyByteBuffer() {
+    ByteBuffer buffer = mBuffer.asReadOnlyBuffer();
+    buffer.position(0);
+    return buffer;
+  }
+
+  /**
+  * Deallocate the Netty ByteBuf if we used ByteBuf to construct this DataByteBuffer.
```
Comment (gpang) [inline common/src/main/java/tachyon/network/protocol/databuffer/DataNettyBuffer.java]: Technically, I don't think it is deallocating the buffer, but just reducing the reference count so netty can deallocate or reuse it later.

File: common/src/main/java/tachyon/network/protocol/databuffer/DataNettyBuffer.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  *
+  * @return True if the netty ByteBuf is deallocated or not constructed using ByteBuf.
+  *         As the Netty channel is responsible for performing another {@link ByteBuf#release()},
+  *         this method can return false in unit tests.
+  */
+  public boolean releaseNettyBuffer() {
```
Comment (gpang) [inline common/src/main/java/tachyon/network/protocol/databuffer/DataNettyBuffer.java]: Related to another comment elsewhere, I think DataBuffer should just have a release() in the interface.

File: servers/src/main/java/tachyon/worker/SpaceCounter.java [hunk 8]
```diff
@@ -1,96 +0,0 @@
...
- * or implied. See the License for the specific language governing permissions and limitations under
- * the License.
- */
-
-package tachyon.worker;
-
-/**
- * The worker space counter, in charge of counting and granting spaces in a worker daemon.
- */
-public class SpaceCounter {
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/SpaceCounter.java]: Why is this file removed?

Comment (ooq) [inline servers/src/main/java/tachyon/worker/SpaceCounter.java]: This is the result of merging the updated master; although I don't understand why the diff shows up. The current master doesn't have this file.

**Comment by gpang** [comment]
Thanks for looking into this issue, @ooq !

Are there tests possible for this?

Also, what are the performance implications for this change?

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockInStream.java (hunk 1 above)
Comment (ooq) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockInStream.java]: Oh yes; this is wrong.

**Comment by ooq** [comment]
@gpang Thanks for the comments. Will probably fix the issues later today.

**Comment by apc999** [comment]
@ooq, this PR does not seem to have the right commits as it contains commits from others that have already been merged into master. It makes this PR looks larger than it really is. Could you clean up this PR first so this PR only has code you need?
